Remove depreciated shared settings hook

louwie17 opened the pull request:
Fixes #6900 

Removes the use of the depreciated `woocommerce_shared_settings` hook.

Some of the main changes:
- Add `WCAdminSharedSettings` class to handle new blocks setting addition, created a new filter for this - `woocommerce_admin_shared_settings`
- Removed one unused option `commentModeration` was removed in #3459
- Made use of the `getSetting` function wherever we were using `window.wcSettings` directly (if possible)
- Use `Fragment` instead of `DropZoneWrapper` for wp version 5.8 and up, given it is depreciated.

**A potential extra change?**
[woocommerce-blocks preloads some settings](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/Assets/AssetDataRegistry.php#L76-L90), some of which we also load (listed below), should I remove these on our end? They pretty well look the same.
- [`currency`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L1022)
- [`homeUrl`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L1084)
- [`locale`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L1023-L1032)
- [`orderStatuses`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L1020)
- [`wcVersion`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L1081)

### Screenshots

### Detailed test instructions:

- Smoke test wc-admin, onboarding wizard, analytics pages, marketing page
     - shipping label banner requires jetpack (run a build and `./bin/make-zip woocommerce-admin.zip` and test on Jurassic)
- Run the E2E tests or check the Github action and see if there are no depreciation logs displayed (you should be able to see the depreciated warnings in the E2E test runs of other PRs - https://github.com/woocommerce/woocommerce-admin/actions/workflows/e2e.yml).
Example of the warning that shouldn't appear anymore:
<img width="1287" alt="Screen Shot 2021-08-11 at 10 03 39 AM" src="https://user-images.githubusercontent.com/2240960/129033720-91a2b4e2-6c45-4862-8f35-940259400fee.png">

- Check if theme upload still works on Wordpress 5.7/5.6

<!--
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note using the following CLI command `npm run changelogger -- add` and commit changes.

louwie17 committed:
Initial set up for the WCAdminSettingsRegistry

louwie17 committed:
Make use of filter for admin settings

louwie17 committed:
Add wc/settings package and remove depreciated dropzone wrapper

louwie17 committed:
Make use of the getSetting function instead of accessing window directly

louwie17 committed:
Add comments

louwie17 committed:
Add changelog

louwie17 commented on packages/wc-admin-settings/src/index.js:
```diff
@@ -5,13 +5,18 @@ import { __ } from '@wordpress/i18n';
+	if ( ! mutableSources.includes( key ) ) {
+		SOURCE[ key ] = settings.admin[ key ];
+	}
+} );
```
Include the `admin` namespace settings

louwie17 commented on client/profile-wizard/steps/theme/uploader.js:
```diff
@@ -68,10 +69,13 @@ class ThemeUploader extends Component {
 		} );
+		const DropZoneWrapper = isWpVersion( '5.8', '>=' )
+			? Fragment
+			: DropZoneProvider;
```
Use a `Fragment` if php version is 5.8 or above.

joshuatf replied:
I assume this means if the WP version is 5.8 or above, right?

Does `Fragment` not work with older versions?

We do have some styling issues here while uploading.  Not entirely sure if they were introduced in this PR or pre-existing.

<img width="384" alt="Screen Shot 2021-08-17 at 2 50 47 PM" src="https://user-images.githubusercontent.com/10561050/129783903-25147276-9866-4503-a991-7d02ab7db574.png">

louwie17 replied:
Let me check, and I will add a comment for the logic. It is the `DropZoneProvider` that is depreciated in WP version 5.8 and up, I am using `Fragment` as the replacement for the time being.

louwie17 replied:
As mentioned below, the styling end up breaking because the `DropzoneProvider` has been removed which the css relied up on.

louwie17 commented on package.json:
```diff
@@ -92,6 +92,7 @@
 		"@automattic/explat-client-react-helpers": "0.0.2",
 		"@woocommerce/e2e-environment": "0.2.2",
 		"@woocommerce/e2e-utils": "0.1.5",
+		"@woocommerce/settings": "^1.0.0",
```
Need for the `isWpVersion` function

joshuatf replied:
Weird that this function is in `@woocomerce/settings` 🤔

louwie17 replied:
I thought so to, maybe I will have another look to see if it's available anywhere else.

louwie17 committed:
Remove un needed changes

joshuatf commented on src/WCAdminSharedSettings.php:
```diff
@@ -0,0 +1,68 @@
+		if ( did_action( 'woocommerce_blocks_loaded' ) ) {
+			$this->on_plugins_loaded();
+		} else {
+			add_action( 'woocommerce_blocks_loaded', array( $this, 'on_woocommerce_blocks_loaded' ), 10 );
```
Nice 💯

joshuatf reviewed (commented):
Very nice fix here, @louwie17!

Should `window.wcSettings.preloadOptions` also be moved under the `admin` namespace?

Looking good overall.  Just left a couple comments in the code.

louwie17 committed:
Update dropzone provider logic and fix styling

louwie17 commented:
> Should `window.wcSettings.preloadOptions` also be moved under the `admin` namespace?

Nice catch! I missed that one, they are included under the `admin` namespace and I guess I missed updating this one case.

@joshuatf it should be good for a re-review, I also fixed the uploader styling, the reason it broke was because of the DropzoneProvider being depreciated (prob only a wp 5.8 issue). The css relied on `components-drop-zone__provider` which didn't exist anymore.
As of now I have left that in, but changed the wrapper logic a bit and added a `woocommerce-theme-uploader__dropzone-wrapper` instead.

louwie17 committed:
Remove old css

joshuatf commented on client/profile-wizard/steps/theme/uploader.js:
```diff
@@ -17,6 +17,27 @@ import { noop } from 'lodash';
+ * @param {Node} [props.children] Children of react component.
+ * @param {string} [props.className] Additional class name to style the component.
+ */
+const DropZoneWrapper = ( { children, className } ) => {
```
Nice wrapper!

joshuatf commented on client/profile-wizard/steps/theme/uploader.js:
```diff
@@ -17,6 +17,27 @@ import { noop } from 'lodash';
+ * @param {string} [props.className] Additional class name to style the component.
+ */
+const DropZoneWrapper = ( { children, className } ) => {
+	const isDropzoneProviderDepreciated = isWpVersion( '5.8', '>=' );
```
```suggestion
	const isDropzoneProviderDeprecated = isWpVersion( '5.8', '>=' );
```

joshuatf commented on client/profile-wizard/steps/theme/uploader.js:
```diff
@@ -17,6 +17,27 @@ import { noop } from 'lodash';
+const DropZoneWrapper = ( { children, className } ) => {
+	const isDropzoneProviderDepreciated = isWpVersion( '5.8', '>=' );
+
+	if ( isDropzoneProviderDepreciated ) {
```
```suggestion
	if ( isDropzoneProviderDeprecated ) {
```

joshuatf reviewed (approved):
Thanks for the changes!  This is testing well for me and LGTM pending a couple minor typos! 🎉 

>woocommerce-blocks preloads some settings, some of which we also load (listed below), should I remove these on our end? They pretty well look the same.

I'm hesitant to rely on another package's preloaded settings, but that fear may not be rational.  Can we open an issue to look into this in a follow-up?

louwie17 commented:
> I'm hesitant to rely on another package's preloaded settings, but that fear may not be rational. Can we open an issue to look into this in a follow-up?

@joshuatf I was actually of the same opinion, but kinda stuck with it as we were using it prior and `woocommerce-blocks` is pretty integrated into `woocommerce`. That being said, given the amount of work I did in this PR, it would take pretty minimal work to have our own `window.wcAdminSettings`. Which I would totally be on-board with, thoughts?
